### PR TITLE
BUG: avoid triggering numpy deprecation warning in assert functions for nested array with empty array/list

### LINKED ISSUE #59776: BUG: assert_frame_equal triggers DeprecationWarning with column with empty lists/arrays
We are running into this in the pyarrow test suite. Reproducer on current main:

```python
empty_list_array1 = np.empty((3,), dtype=object)
empty_list_array1.fill([])
df1 = pd.DataFrame({'a': empty_list_array1})

empty_list_array2 = np.empty((3,), dtype=object)
empty_list_array2.fill(np.array([], dtype=object))
df2 = pd.DataFrame({'a': empty_list_array2})

import warnings
warnings.simplefilter("always")
```

gives:

```
>>> pd.testing.assert_frame_equal(df1, df2)
/pandas/core/dtypes/missing.py:503: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  return lib.array_equivalent_object(left, right)
```

The above is with numpy 1.26 and pandas main, but also see the same with latest numpy and pandas (2.1.1 and 2.2.2).

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
I haven't added a whatsnew note, because it shouldn't change behaviour but just avoids a warning being triggered (so thought that to not warrant a mention in the release notes)

- [x] closes #59776
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -600,6 +600,8 @@ def array_equivalent_object(ndarray left, ndarray right) -> bool:
                     if not array_equivalent(x, y):
                         return False
 
+            elif PyArray_Check(x) or PyArray_Check(y):
+                return False
```
Comment (jorisvandenbossche) [inline pandas/_libs/lib.pyx:604]: Below we would essentially do `bool(left == right)` which raises an error for arrays/lists of size > 1 (and that error would then be caught below, turned into `return False`. So for now, just directly returning False here when left or right is a numpy array and the other not.

That then avoids doing this `bool(left == right)` for empty array/list case, which triggers the warning.

Comment (jorisvandenbossche) [reply pandas/_libs/lib.pyx:604]: The first `if` just before is `if PyArray_Check(x) and PyArray_Check(y)`, so the `or` is sufficient

**Review by WillAyd — APPROVED** [review]

**Comment by lithomas1** [comment]
Does this need a backport?

**Comment by jorisvandenbossche** [comment]
No, it's just for silencing a warning in a corner case, not worth of backporting

**Pull request merged**